terracotta/handlers/metadata.py

```diff
@@ -3,18 +3,52 @@
 Handle /metadata API endpoint.
 """
 
-from typing import Mapping, Sequence, Dict, Any, Union
+from typing import Mapping, Sequence, Dict, Any, Union, List, Optional
 from collections import OrderedDict
 
 from terracotta import get_settings, get_driver
 from terracotta.profile import trace
 
 
+def filter_metadata(
+    metadata: Dict[str, Any], columns: Optional[List[str]]
+) -> Dict[str, Any]:
+    """Filter metadata by columns, if given"""
+    assert (
+        columns is None or len(columns) > 0
+    ), "columns must either be a non-empty list or None"
+
+    if columns:
+        metadata = {c: metadata[c] for c in columns}
+
+    return metadata
+
+
 @trace("metadata_handler")
-def metadata(keys: Union[Sequence[str], Mapping[str, str]]) -> Dict[str, Any]:
+def metadata(
+    columns: Optional[List[str]], keys: Union[Sequence[str], Mapping[str, str]]
+) -> Dict[str, Any]:
     """Returns all metadata for a single dataset"""
     settings = get_settings()
     driver = get_driver(settings.DRIVER_PATH, provider=settings.DRIVER_PROVIDER)
-    metadata = driver.get_metadata(keys)
+    metadata = filter_metadata(driver.get_metadata(keys), columns)
     metadata["keys"] = OrderedDict(zip(driver.key_names, keys))
     return metadata
+
+
+@trace("multiple_metadata_handler")
+def multiple_metadata(
+    columns: Optional[List[str]], datasets: List[List[str]]
+) -> List[Dict[str, Any]]:
+    """Returns all metadata for multiple datasets"""
+    settings = get_settings()
+    driver = get_driver(settings.DRIVER_PATH, provider=settings.DRIVER_PROVIDER)
+    key_names = driver.key_names
+
+    out = []
+    for dataset in datasets:
+        metadata = filter_metadata(driver.get_metadata(dataset), columns)
+        metadata["keys"] = OrderedDict(zip(key_names, dataset))
+        out.append(metadata)
```

Comment:
There's a `with driver.connect()` missing around this block. Without it, the connection is opened and closed for every dataset which should have a decisive performance impact.

Reply:
Great catch - didn't know - thanks 😄

Reply:
Did this solve the performance issues? No need to benchmark it in production, the test you had should suffice.